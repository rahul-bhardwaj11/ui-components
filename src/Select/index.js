import React, { Component } from 'react';
import PropTypes from 'prop-types';
import 'antd/lib/dropdown/style/css';
import AntSelect from 'antd/lib/select';
import 'antd/lib/select/style/index.css';
import styled from 'styled-components';
import selectArrowIcon from '../styles/icons/downFillcaret.svg';
import StringToHTML from '../StringToHTML';
import classnames from 'classnames';
import theme from '../styles/theme';
import Icon from '../Icon';
const Option = AntSelect.Option;
import ReactDOM from 'react-dom';

const MtWrapper = styled.span`
  .selectDropdownStyle {
    z-index: 999999;
    .ant-select-dropdown-menu-item {
     div {
         width: ${props => (props.showTick ? '95%' : null)};
         text-overflow: ellipsis;
         overflow: hidden;
         white-space: nowrap;
       }
     }
   }

  .ant-select-dropdown-menu-item-disabled{
    opacity:0.3;
    pointer-events: none;
  }

  .icon-tick {
    display: none;
  }
  .ant-select, .ant-select-dropdown{
    font-family: inherit;
  }
  .ant-select-arrow {
    display: block;
    background: url('${selectArrowIcon}') no-repeat 1px 4px;
    background-size: 8px;
    width: 9px;
    height: 9px;
    &::before {
      content: '';
    }
    svg{
      display:none;
    }
  }
  .ant-select-selection {
    &.ant-select-selection--single {
      border: 1px solid ${theme.colors.ALTO};
      &:hover {
        border: 1px solid ${theme.colors.SILVER};
      }
      &:focus,
      &:active {
        box-shadow: none;
        border: 1px solid ${theme.colors.SILVER};
      }
      &.disabled {
        background-color: #f8f8f8;
        border: 1px solid #e8e8e8;
      }
    }
    .ant-select-open {
      .ant-select-selection {
        border: 1px solid ${theme.colors.SILVER};
      }
    }
  }

  .ant-select-open {
    .ant-select-selection {
      &.ant-select-selection--multiple {
        border: 1px solid transparent;
        outline: none;
        &:hover,
        &:focus,
        &:active {
          border: 0;
          border-color: transparent;
          outline: none;
          box-shadow: none;
        }
      }
    }
  }
  .ant-select-dropdown-menu {
    background-color: ${theme.colors.WHITE};
    color: ${theme.colors.GREY};
    margin: 8px;
    max-height: 235px;

    .ant-select-dropdown-menu-item {
      border-radius: 4px;
      margin-bottom: 4px;
      &:last-child{
        margin-bottom: 0px;
      }
      color: ${theme.colors.DARK_OUTER_SPACE};
      &:hover {
        background-color: ${theme.colors.INDIGO};
        color: ${theme.colors.WHITE};
        border-radius: 4px;
      }
    }
    .ant-select-dropdown-menu-item-active {
      background-color: ${theme.colors.WHITE};
      color: ${theme.colors.SHARK};
    }

  .ant-select-dropdown-menu-item-selected{
       background-color:${theme.colors.TROPICAL_BLUE};
      .icon-tick {
        content: '';
        font-size: 7px;
        font-weight: bold;
        position: absolute;
        display: block;
        right: 5px;
        top: 13px;
      }
      &:hover {
        .icon-tick {
          display: block;
        }
      }
    }
  }
  .ant-select-selection-selected-value {
    font-weight: normal;
    div {
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
      max-width: 100%;
    }
  }
  .ant-select-selection--multiple {
    width: 688px;
    padding: 5px;
    border: 0px;
    &:focus {
      border-color: transparent;
      box-shadow: none;
    }
    .ant-select-selection__rendered {
      & > ul {
        & > .ant-select-selection__choice {
          border-radius: 14px;
          border: 1px solid ${theme.colors.ALTO};
          border-radius: 16px;
          background-color: ${theme.colors.PORCELAIN};
          color: ${theme.colors.OUTER_SPACE};
          font-size: 12px;
          .ant-select-selection__choice__content {
            margin-right: 5px;
          }
          &:hover {
            border: 1px solid #468ee5;
            background-color: ${theme.colors.WHITE};
            color: #468ee5;
            cursor: pointer;
            .ant-select-selection__choice__remove {
              color: #468ee5;
            }
          }
          .ant-select-selection__choice__remove {
            color: ${theme.colors.OUTER_SPACE};
            font-size: 16px;
          }
        }
      }

      & > ul > li {
        margin-bottom: 3px;
      }
    }
  }
  .ant-select-dropdown--multiple {
    z-index: 999999;
  }
`;

class Select extends Component {
  static propTypes = {
    options: PropTypes.any.isRequired,
    onChange: PropTypes.func,
    style: PropTypes.object,
    defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    title: PropTypes.string,
    className: PropTypes.string,
    getPopupContainer: PropTypes.func,
    showTick: PropTypes.bool,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.array
    ])
  };

  static defaultProps = {
    style: { minWidth: 125 },
    showTick: true
  };

  element = null;
  constructor(p) {
    super(p);
    this.element = document.createElement('div');
    this.selectRef = React.createRef();
  }

  state = {
    key: 'select'
  };

  componentDidMount() {
    document.body.appendChild(this.element);
  }

  componentWillUnmount() {
    document.body.removeChild(this.element);
  }

  // hack for controlled value
  componentWillReceiveProps(nextProps) {
    if (nextProps.value != this.props.value) {
      this.setState({ key: nextProps.value });
    }
  }

  render() {
    let { options, style, showTick } = this.props;
    const container =
      this.props.getPopupContainer && this.props.getPopupContainer();
    return (
      <React.Fragment>
        {ReactDOM.createPortal(
          <MtWrapper style={style} innerRef={this.selectRef} />,
          container || this.element
        )}
        <MtWrapper style={style} key={this.state.key} showTick={showTick}>
          <AntSelect
            {...this.props}
            onClick={event => {
              event.stopPropagation();
            }}
            getPopupContainer={() => {
              return this.selectRef.current;
            }}
            dropdownClassName={classnames(
              'selectDropdownStyle',
              this.props.className
            )}
          >
            {options.map(option => {
              return (
                <Option
                  key={option.key}
                  value={option.key}
                  title={this.props.title || option.title}
                >
                  {typeof option.content === 'string' ? (
                    <StringToHTML content={option.content} />
                  ) : (
                    option.content
                  )}
                  {showTick && <Icon type="tick" />}
                </Option>
              );
            })}
          </AntSelect>
        </MtWrapper>
      </React.Fragment>
    );
  }
}

class AsyncSelect extends Component {
  static propTypes = {
    handleSearch: PropTypes.func.isRequired,
    handleChange: PropTypes.func.isRequired,
    options: PropTypes.array,
    style: PropTypes.object,
    placeholder: PropTypes.string,
    title: PropTypes.string,
    value: PropTypes.array
  };

  static defaultProps = {
    style: { minWidth: 125 }
  };

  handleSearch = async value => {
    this.props.handleSearch(value);
  };

  handleChange = value => {
    this.props.handleChange(value);
  };

  render() {
    const { style } = this.props;

    return (
      <MtWrapper
        innerRef={el => {
          if (el) {
            this.selectDropdownRef = el;
          }
        }}
        style={style}
      >
        <AntSelect
          {...this.props}
          showSearch
          value={this.props.value}
          onSearch={this.handleSearch}
          onChange={this.handleChange}
          filterOption={false}
          defaultActiveFirstOption={false}
          showArrow={false}
        >
          {this.props.options &&
            this.props.options.map(option => {
              return (
                <Option
                  key={option.key}
                  value={option.key}
                  title={this.props.title || option.title}
                  disabled={option.disabled}
                >
                  {typeof option.content === 'string' ? (
                    <StringToHTML content={option.content} />
                  ) : (
                    option.content
                  )}
                </Option>
              );
            })}
        </AntSelect>
      </MtWrapper>
    );
  }
}

Select.Async = AsyncSelect;

export default Select;
