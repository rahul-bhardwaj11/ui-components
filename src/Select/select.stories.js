import React from 'react';
import { storiesOf } from '@storybook/react';
import Select from './index';
import { withInfo } from '@storybook/addon-info';
import { withKnobs, object } from '@storybook/addon-knobs';

const stories = storiesOf('Select', module);
stories.addDecorator(withKnobs);

const options = [
  { key: '1', content: 'First Item' },
  { key: '2', content: 'Second Item' },
  { key: '3', content: 'Third Item', disabled: true }
];

class ControlledSelect extends React.Component {
  state = {
    value: '3'
  };
  render() {
    return (
      <Select
        options={object('options', options)}
        value={this.state.value}
        onSelect={key => {
          this.setState({ value: key });
        }}
      />
    );
  }
}
stories.add(
  'Default Select',
  withInfo('Basic usage of the Select')(() => (
    <Select
      options={object('options', options)}
      defaultValue="Select"
      showTick={true}
    />
  ))
);

stories
  .add(
    'Controlled Select',
    withInfo('Basic usage of the Select')(() => <ControlledSelect />)
  )
  .add(
    'Multiple Select',
    withInfo('Basic usage of the Multiple Select')(() => (
      <Select
        options={object('options', options)}
        defaultValue="Select"
        mode="multiple"
      />
    ))
  )
  .add(
    'Async Select',
    withInfo('Basic usage of Async Select')(() => (
      <Select.Async
        mode="multiple"
        placeholder="Search any artist"
        handleSearch={handleSearch}
        handleChange={console.log} //eslint-disable-line
        notFoundContent={null}
      />
    ))
  );
stories.add(
  'Select with different dropdown width',
  withInfo('Basic usage of the Select')(() => (
    <Select
      options={object('options', options)}
      defaultValue="Select"
      dropdownMatchSelectWidth={false}
      dropdownStyle={{ width: '150px' }}
    />
  ))
);

function handleSearch(value) {
  const url = 'https://itunes.apple.com/search?term=';
  return fetch(`${url}${value}`)
    .then(response => response.json())
    .then(({ results }) =>
      results.map(v => ({
        content: `${v.trackName} by ${v.artistName}`,
        key: `${v.trackId}-${v.artistId}`
      }))
    );
}
