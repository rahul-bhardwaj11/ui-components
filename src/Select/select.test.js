import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Select from './index';

describe('Select', () => {
  const options = [
    { key: '1', content: 'First Item' },
    { key: '2', content: 'Second Item' },
    { key: '3', content: 'Third Item' }
  ];
  it('matches default snapshot', () => {
    const tree = renderer.create(<Select options={options} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  const changeSpy = jest.fn();
  const select = shallow(<Select options={options} onChange={changeSpy} />);
  it('should handle click', () => {
    const stopPropagation = jest.fn();
    select.find('Select').simulate('click', { stopPropagation });
    expect(stopPropagation).toHaveBeenCalled();
  });

  it('should handle change', () => {
    select.find('Select').simulate('change');
    expect(changeSpy).toHaveBeenCalled();
  });
});
