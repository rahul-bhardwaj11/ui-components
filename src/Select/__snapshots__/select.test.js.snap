// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select matches default snapshot 1`] = `
<div
  className="ant-select ant-select-enabled"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={undefined}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={undefined}
  onMouseLeave={undefined}
  onTouchStart={[Function]}
  style={
    Object {
      "minWidth": 120,
    }
  }
>
  <div
    aria-autocomplete="list"
    aria-expanded={false}
    aria-haspopup="true"
    className="ant-select-selection
            ant-select-selection--single"
    onKeyDown={[Function]}
    role="combobox"
    tabIndex={0}
  >
    <div
      className="ant-select-selection__rendered"
    />
    <span
      className="ant-select-arrow"
      onClick={[Function]}
      style={
        Object {
          "WebkitUserSelect": "none",
          "userSelect": "none",
        }
      }
      unselectable="on"
    >
      <b />
    </span>
  </div>
</div>
`;
